### PR TITLE
Project Euler problem148

### PR DESCRIPTION
**Pull request by saitejamanchi**
### Describe your change:

Added solution for Problem 148 in Project Euler.

* [X] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### Checklist:
* [X] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [X] This pull request is all my own work -- I have not plagiarized.
* [X] I know that pull requests will not be merged if they fail the automated tests.
* [X] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [ ] All new Python files are placed inside an existing directory.
* [X] All filenames are in all lowercase characters with no spaces or dashes.
* [X] All functions and variable names follow Python naming conventions.
* [X] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [X] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [X] All new algorithms include at least one URL that points to Wikipedia or another similar explanation.
* [ ] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

File: project_euler/problem_148/sol1.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+We only need to generate half of the values.
+We then count the values which are not divisible by 7.
+We only store the remainders(when divided by 7) in the list to reduce memory usage.
+
+Note: In the original problem, we need to calculate for 10^9 rows
+      but we took 10^4 rows here by default.
+"""
+
+
+def solution(pascal_row_count: int = 10**4) -> int:
```
Comment (rohan472000) [inline project_euler/problem_148/sol1.py:31]: try this:

`def solution(pascal_row_count: int = 10**9) -> int:`

Comment (saitejamanchi) [reply project_euler/problem_148/sol1.py:31]: Hey,

I've tried with 10**9 but one of the tests failed due to timeout as it was taking more than 6 hours to run.

Comment (rohan472000) [reply project_euler/problem_148/sol1.py]: It's not uncommon for brute force solutions to take a long time or run into timeout issues when dealing with very large inputs. In this case, the input value of 10**9 is extremely large, and the current implementation of the "solution" function does not have an efficient algorithm to handle such large inputs.

**Review by rohan472000 — CHANGES_REQUESTED** [review]
kindly go through below code, it should fix your timeout issue.

"""def solution(pascal_row_count: int = 10**9) -> int:
    """
    To evaluate the solution, use solution()
    >>> solution(3)
    6
    >>> solution(10)
    40
    >>> solution(100)
    2361
    """
    # Helper function to compute the number of odd binomial coefficients in a row
    def odd_binomials(n: int) -> int:
        res = 0
        while n > 0:
            res += n % 2
            n //= 2
        return res

    # Compute the sum of odd binomial coefficients modulo 7 for each row up to pascal_row_count
    result = 0
    for k in range(pascal_row_count):
        result += odd_binomials(k) % 7

    return result
"""

File: project_euler/problem_148/sol1.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+We then count the values which are not divisible by 7.
+We only store the remainders(when divided by 7) in the list to reduce memory usage.
+
+Note: In the original problem, we need to calculate for 10^9 rows
+      but we took 10^4 rows here by default.
+"""
+
+
+def solution(pascal_row_count: int = 10**9) -> int:
+    """
```
Comment (rohan472000) [inline project_euler/problem_148/sol1.py:32]: use below codes for better efficiency which will solve the timeout issue:-

    

      def solution(pascal_row_count: int = 10**9) -> int:
          """
          To evaluate the solution, use solution()
          >>> solution(3)
          6
          >>> solution(10)
          40
          >>> solution(100)
          2361
          """
      
          # Helper function to compute the number of odd binomial coefficients in a row
          def odd_binomials(n: int) -> int:
              res = 0
              while n > 0:
                  res += n % 2
                  n //= 2
              return res
      
          # Compute the sum of odd binomial coefficients modulo 7 for each row up to pascal_row_count
          result = 0
          for k in range(pascal_row_count):
              result += odd_binomials(k) % 7
      
          return result

**Review by rohan472000 — CHANGES_REQUESTED** [review]

File: project_euler/problem_148/sol1.py (hunk 2 above)
Comment (rohan472000) [reply project_euler/problem_148/sol1.py:32]: also add docs in `odd_binomials`, you can put above comment of that function inside as a doc.

**Comment by rohan472000** [comment]
use `ruff .` then `ruff . --fix` to fix ruff issues.

File: project_euler/problem_148/sol1.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+      1          5         10         10          5          1
+1          6         15         20         15          6          1
+However, if we check the first one hundred rows, we will find that
+only 2361 of the 5050 entries are not divisible by 7.
+Find the number of entries which are not divisible by 7
+in the first one billion (109) rows of Pascal's triangle.
+"""
+
+
+def get_num_binomials(row_num: int) -> int:
```
[the remark this answers is not in the document]
Comment (rohan472000) [reply project_euler/problem_148/sol1.py:21]: @saitejamanchi add doctests.

[the remark this answers is not in the document]
Comment (saitejamanchi) [reply project_euler/problem_148/sol1.py:21]: @rohan472000 Sure, will add.

Thanks a lot for the help!!

**Comment by rohan472000** [comment]
kindly fix ruff issue before committing...by `ruff .` then `ruff , --fix`

**Comment by saitejamanchi** [comment]
Hi @dhruvmanila ,

Could you please look into this pull request?

**Pull request closed without merging**